### PR TITLE
[pack] extension bundle probing and download

### LINKED ISSUE #3711: Add logic for locating bundles at probing locations

### LINKED ISSUE #3961: Download the extension bundle if it is not present at the default location
If the extension bundle is not present in the default location, then download the extension bundle if persistent storage is available

### PR DESCRIPTION
**Pull request by soninaren**
Resolves #3961 
Resolves #3711

### CODE REVIEW

**Comment by pragnagopa** [comment]
@soninaren / @SatishRanjan - Coincidentally we have two features which require probing directories for dependencies.  It makes sense to define a common interface for both of them. @fabiocav -  thoughts?

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    internal class ExtensionBundleManager : IExtensionBundleManager
+    {
+        private readonly IEnvironment _environment;
+        private readonly IOptions<ExtensionBundleOptions> _options;
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        public ExtensionBundleManager(IEnvironment environment,
+                                      IOptions<ExtensionBundleOptions> extensionBundleOptions,
+                                      HttpClient httpClient,
```
Comment (fabiocav) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs]: Given the usage pattern, this could be just managed within the scope of the download method.

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            bundlePath = null;
+            string version = null;
+            var paths = new List<string>(_options.Value.ProbingPaths)
+            {
+                _options.Value.DownloadPath
+            };
+
+            for (int i = 0; i < paths.Count; i++)
+            {
+                var path = paths.ElementAt(i);
```
Comment (fabiocav) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs]: May want to use the indexer here

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs [hunk 3]
```diff
@@ -53,5 +63,31 @@ private static void ValidateBundleId(string id)
...
             }
         }
+
+        public void ConfigureProbingPaths(ExtensionBundleOptions options)
+        {
+            options.ProbingPaths = new List<string>();
+
+            if (_environment.IsAppServiceWindowsEnvironment())
+            {
+                string windowsDefaultPath = Path.Combine(Environment.ExpandEnvironmentVariables(ScriptConstants.ProgramFilesX86),
```
Comment (fabiocav) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs]: Use https://docs.microsoft.com/en-us/dotnet/api/system.environment.getfolderpath?view=netframework-4.7.2 instead

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs [hunk 4]
```diff
@@ -53,5 +63,31 @@ private static void ValidateBundleId(string id)
...
+                                                         ScriptConstants.DefaultExtensionBundleDirectory,
+                                                         options.Id);
+
+                options.ProbingPaths.Add(windowsDefaultPath);
+            }
+
+            if (_environment.IsLinuxAppServiceEnvironment())
+            {
+                string linuxDefaultPath = Path.Combine("/", ScriptConstants.DefaultExtensionBundleDirectory, options.Id);
```
Comment (fabiocav) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs]: Use the PathSeparator here insted of '/'

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs [hunk 5]
```diff
@@ -53,5 +63,31 @@ private static void ValidateBundleId(string id)
...
+
+                string deploymentPackageBundlePath = Path.Combine(
+                    Environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsiteHomePath),
```
Comment (fabiocav) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs]: Same comment about `Environment` use here

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                if (FileUtility.DirectoryExists(path))
+                {
+                    var bundleDirectories = FileUtility.EnumerateDirectories(path);
+                    version = FindBestVersionMatch(_options.Value.Version, bundleDirectories);
+                }
+
+                if (!string.IsNullOrEmpty(version))
+                {
+                    bundlePath = Path.Combine(path, version);
+                    _logger.LogInformation(Resources.ExtensionBundleFound);
```
Comment (fabiocav) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleManager.cs]: Would be nice to log the version as well, so we know which bundle was actually used.

File: src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs [hunk 7]
```diff
@@ -31,13 +34,20 @@ public void Configure(ExtensionBundleOptions options)
...
+                if (_environment.IsAppServiceEnvironment())
+                {
+                    options.DownloadPath = Path.Combine(_environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsiteHomePath),
+                                                   "data", "Functions", ScriptConstants.ExtensionBundleDirectory, options.Id);
+                    ConfigureProbingPaths(options);
+                }
             }
         }
 
         private static void ConfigureBundleVersion(IConfigurationSection configurationSection, ExtensionBundleOptions options)
```
Comment (pragnagopa) [inline src/WebJobs.Script/BindingExtensionBundle/ExtensionBundleOptionsSetup.cs]: Change to private. Remove static

File: src/WebJobs.Script/BindingExtensionBundle/IExtensionBundleManager.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Script.BindingExtensionBundle
+{
+    internal interface IExtensionBundleManager
+    {
```
Comment (pragnagopa) [inline src/WebJobs.Script/BindingExtensionBundle/IExtensionBundleManager.cs]: Remove if not needed

File: src/WebJobs.Script/Properties/Resources.resx [hunk 9]
```diff
@@ -132,4 +132,34 @@
   <data name="ExtensionBundleConfigMissingVersion" xml:space="preserve">
     <value>The value of version property in extensionBundle section of {0} file is invalid or missing. See https://aka.ms/functions-hostjson for more information</value>
   </data>
+  <data name="DownloadComplete" xml:space="preserve">
+    <value>Completed downloading {0} to {1}</value>
```
[the remark this answers is not in the document]
Comment (soninaren) [reply src/WebJobs.Script/Properties/Resources.resx]: This is when the extension bundle zip completes downloading.

File: src/WebJobs.Script/Config/ExtensionBundleOptions.cs [hunk 10]
```diff
@@ -2,14 +2,21 @@
...
     {
         public string Id { get; set; }
 
         public VersionRange Version { get; set; }
+
+        public ICollection<string> ProbingPaths { get; set; }
+
+        public string DownloadPath { get; set; }
+
+        public bool EnsureLatest { get; set; }
```
Comment (pragnagopa) [inline src/WebJobs.Script/Config/ExtensionBundleOptions.cs]: What is the purpose of EnsureLatest? should we default to the latest version available in the bundles?

Comment (soninaren) [reply src/WebJobs.Script/Config/ExtensionBundleOptions.cs]: Setting this would get the latest extension bundle within the given version range. Not the absolute latest

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs [hunk 11]
```diff
@@ -48,16 +52,32 @@ private static string[] GetBuiltinExtensionAssemblies()
 
         public Type[] GetStartupTypes()
         {
-            IEnumerable<Type> startupTypes = GetExtensionsStartupTypes();
+            IEnumerable<Type> startupTypes = GetExtensionsStartupTypesAsync().ConfigureAwait(false).GetAwaiter().GetResult();
```
Comment (pragnagopa) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: Changed the mehtod to async but making this a sync call? If possible we should avoid this pattern

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs [hunk 12]
```diff
@@ -48,16 +52,32 @@ private static string[] GetBuiltinExtensionAssemblies()
...
 
             return startupTypes
                 .Distinct(new TypeNameEqualityComparer())
                 .ToArray();
         }
 
-        public IEnumerable<Type> GetExtensionsStartupTypes()
+        public async System.Threading.Tasks.Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
```
Comment (pragnagopa) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: Same as above. Why did you change this to async? I do not see any async calls in this method

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs (hunk 11 above)
Comment (soninaren) [reply src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: It is unavoidable at this place since this is expected by webjobs sdk.

File: src/WebJobs.Script/Config/ExtensionBundleOptions.cs (hunk 10 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Config/ExtensionBundleOptions.cs]: Can you give an example? As users already pick a version range, we should just default to latest version we have downloaded on to the machines.

File: src/WebJobs.Script/Environment/EnvironmentSettingNames.cs [hunk 13]
```diff
@@ -50,5 +50,7 @@ public static class EnvironmentSettingNames
 
         public const string LinuxAzureAppServiceStorage = "WEBSITES_ENABLE_APP_SERVICE_STORAGE";
         public const string CoreToolsEnvironment = "FUNCTIONS_CORETOOLS_ENVIRONMENT";
+
+        public const string AlternateCdnUri = "FUNCTIONS_ALTERNATE_CDN_URI";
```
Comment (pragnagopa) [inline src/WebJobs.Script/Environment/EnvironmentSettingNames.cs]: Why is this called AlternateUri? is this a secondary?

File: src/WebJobs.Script/ScriptConstants.cs [hunk 14]
```diff
@@ -121,6 +121,12 @@ public static class ScriptConstants
 
         public const string AzureWebJobsHostsContainerName = "azure-webjobs-hosts";
 
+        public const string DefaultExtensionBundleDirectory = "FuncExtensionBundles";
+        public const string ExtensionBundleDirectory = "ExtensionBundles";
+        public const string ExtensionBundleCdnUri = "https://functionscdn.azureedge.net/public/extensionBundles";
+        public const string ExtensionBundleMetadatFile = "bundle.json";
+        public const string VersionIndexFile = "index.json";
```
Comment (pragnagopa) [inline src/WebJobs.Script/ScriptConstants.cs]: Follow same naming convention. Prefix with ExtensionBundle.
Change to ExtensionBundleDefaultDirectory, ExtensionBundleVersionIndexFile

File: test/WebJobs.Script.Tests.Integration/Host/StandbyManagerTests.cs [hunk 15]
```diff
@@ -328,10 +329,17 @@ protected async Task InitializeTestHostAsync(string testDirName, IEnvironment en
...
+                    
                     if (source != null)
                     {
                         c.Sources.Remove(source);
                     }
+
+                    var bundleSource = c.Sources.OfType<ExtensionBundleConfigurationSource>().SingleOrDefault();
+                    if (bundleSource != null)
+                    {
+                        c.Sources.Remove(bundleSource);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests.Integration/Host/StandbyManagerTests.cs]: Why is this needed?

File: test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs [hunk 16]
```diff
@@ -9,8 +9,10 @@
 using Microsoft.Azure.WebJobs.Extensions;
 using Microsoft.Azure.WebJobs.Extensions.Http;
 using Microsoft.Azure.WebJobs.Extensions.Storage;
+using Microsoft.Azure.WebJobs.Script.BindingExtensionBundle;
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs]: Nit: Can you please change the name space? ExtensionBundle  sounds more appropriate than BindingExtensionBundle

File: test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs [hunk 17]
```diff
@@ -34,9 +36,84 @@ public void GetExtensionsStartupTypes_FiltersBuiltinExtensions()
...
+                // Assert
+                Assert.Single(types);
+                Assert.Equal(typeof(AzureStorageWebJobsStartup).FullName, types.Single().FullName);
+                Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, $"The extension startup type '{references[0].TypeName}' belongs to a builtin extension")));
+                Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, $"The extension startup type '{references[1].TypeName}' belongs to a builtin extension")));
+            }
+        }
+
+        [Fact]
+        public async System.Threading.Tasks.Task GetExtensionsStartupTypes_ExtensionBundleReturnsNullPath_ReturnsNull()
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs]: Nit: Just use Task

File: test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs [hunk 18]
```diff
@@ -34,9 +36,84 @@ public void GetExtensionsStartupTypes_FiltersBuiltinExtensions()
...
+                new ExtensionReference { Name = "Timers", TypeName = typeof(ExtensionsWebJobsStartup).AssemblyQualifiedName },
+                new ExtensionReference { Name = "Storage", TypeName = typeof(AzureStorageWebJobsStartup).AssemblyQualifiedName },
+            };
+
+            var extensions = new JObject
+            {
+                { "extensions", JArray.FromObject(references) }
+            };
+
+            var extensionBundleManager = new Mock<IExtensionBundleManager>();
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs]: nit: rename to mockExtensionBundleManager

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs (hunk 12 above)
Comment (soninaren) [reply src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: have an async call below `string extensionBundlePath = await _extensionBundleManager.GetExtensionBundle();`

File: src/WebJobs.Script/Environment/EnvironmentSettingNames.cs (hunk 13 above)
Comment (soninaren) [reply src/WebJobs.Script/Environment/EnvironmentSettingNames.cs]: This is to help development and to help us provide secondary cdn link incase the first one is busted for some reason.

**Comment by fabiocav** [comment]
@pragnagopa @brettsam @alrod  working on a review here, but it would be good to have your feedback

File: src/WebJobs.Script/Config/ExtensionBundleOptions.cs (hunk 10 above)
Comment (soninaren) [reply src/WebJobs.Script/Config/ExtensionBundleOptions.cs]: Discussed with @pragnagopa offline, We should use / download the latest version of bundle available to us. App service should already have the latest version before CDN so download wouldn't kick in for App service environment.

File: test/WebJobs.Script.Tests.Integration/Host/StandbyManagerTests.cs (hunk 15 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests.Integration/Host/StandbyManagerTests.cs]: This test requires you to remove sources that rely on script root. When i added a new source it broke this test. I changed it to fix the test and get it back to passing state.

File: src/WebJobs.Script/Config/ExtensionBundleOptions.cs [hunk 19]
```diff
@@ -2,14 +2,21 @@
...
-namespace Microsoft.Azure.WebJobs.Script.BindingExtensionBundle
+namespace Microsoft.Azure.WebJobs.Script.Configuration
 {
     public class ExtensionBundleOptions
     {
         public string Id { get; set; }
 
         public VersionRange Version { get; set; }
+
+        public ICollection<string> ProbingPaths { get; set; }
```
Comment (pragnagopa) [inline src/WebJobs.Script/Config/ExtensionBundleOptions.cs]: Change to private set

File: src/WebJobs.Script/Config/ExtensionBundleOptions.cs (hunk 10 above)
Comment (soninaren) [reply src/WebJobs.Script/Config/ExtensionBundleOptions.cs]: Going back and looking at hosting environments and release process. I am reverting to download extension bundle only when a matching bundle is not already present on vm or user storage.

**Comment by soninaren** [comment]
@fabiocav have addressed the PR comments. Let me know if there are any more changes.

**Comment by soninaren** [comment]
Closing this one and opening this one instead https://github.com/Azure/azure-functions-host/pull/4135. We can reiterate the new PR. This would allow me to create draft PR's that build on top of the branch from the new PR.
